Question: NodeJs support

Hey - does this work with NodeJs as well? 
also, does it support —experimental-websocket-port as well or just the P2P port?

And what’s the role of the Hex Encoded private key in this communication?

> In the meantime, I added some fixes that should fix the resolution error that @escapedcat mentioned. So let me know if that looks good on your end now that you are running `v0.0.11`.

With `v0.0.11` we still get these when using webpack without the special resolver:
```sh
ERROR in ./node_modules/lnmessage/dist/messages/InitMessage.js 3:0-38
Module not found: Error: Can't resolve './BitField' in '/Users/user/alby/node_modules/lnmessage/dist/messages'
Did you mean 'BitField.js'?
BREAKING CHANGE: The request './BitField' failed to resolve only because it was resolved as fully specified
(probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
```

No need to rush this. I believe this is non-blocking for us. We can make Alby work with the special rule for now till the build is sorted out.

--- a/README.md
+++ b/README.md
@@ -1,12 +1,11 @@
 # lnmessage
 
-Talk to Lightning nodes from the Browser or NodeJS(after polyfill) apps.
+Talk to Lightning nodes from the Browser and NodeJS apps.
 
 ## Features
 
 - Connect to a lightning node via a WebSocket connection.
-- Works in the Browser without any polyfilling.
-- Works in NodeJS with a simple polyfill
+- Works in the Browser and Node without any polyfilling.
 - Initialise with a session secret to have a persistent node public key for the browser.
 - Control a Core Lightning node via [Commando](https://lightning.readthedocs.io/lightning-commando.7.html) RPC calls.
 - Automatic handling of ping messages to ensure constant connection to the node.
@@ -82,7 +81,7 @@ type LnWebSocketOptions = {
    */
   wsProxy?: string
   /**
-   * When connecting directly to a node, the protocol to use. Defaults to 'wss://'
+   * When connecting directly to a node and not using a proxy, the protocol to use. Defaults to 'wss://'
    */
   wsProtocol?: 'ws:' | 'wss:'
   /**
@@ -230,50 +229,15 @@ class Lnmessage {
 }
 ```
 
-## NodeJS Polyfill
-
-Lnmessage is designed for the browser, but can be adapted to work in NodeJS apps with a simple polyfill:
-
-1. Install `ws` dependency
-
-**Yarn**
-`yarn add ws`
-`yarn add -D @types/ws`
-
-**NPM**
-
-`npm i ws`
-`npm install --save-dev @types/ws`
-
-2. Create a `polyfills.ts` file
-
-```typescript
-import WebSocket from 'ws'
-import crypto from 'crypto'
-
-if (!(<any>global).crypto) {
-  ;(<any>global).crypto = crypto
-}
-
-if (!(<any>global).WebSocket) {
-  ;(<any>global).WebSocket = WebSocket
-}
-```
-
-3. Include polyfills.ts file in your tsconfig.json's includes section.
-4. Import and initialise polyfills.ts at the start of your project.
-
 ## WebSocket Proxy
 
 There are some limitations to connecting to Lightning nodes within a browser. Core Lightning nodes can be directly connected to if the [`experimental-websocket-port`](https://lightning.readthedocs.io/lightningd-config.5.html#experimental-options) option is set in the config. This will allow a direct connection to the node, but if you are running a browser app on https, then it will not allow a connection to a non SSL WebSocket endpoint, so you would need to setup SSL for your node. As far as I know LND nodes do not accept connections via WebSocket at this time.
 
 So to simplify connecting to any Lightning node, you can go through a WebSocket proxy (see [Clams](https://github.com/clams-tech/lnsocket-proxy) and [jb55](https://github.com/jb55/ln-ws-proxy)'s WebSocket proxy server repos). Going through a proxy like this requires no trust in the server. The WebSocket connection is initated with the proxy, which then creates a regular TCP socket connection to the node. Then all messages are fully encrypted via the noise protocol, so the server only sees encrypted binary traffic that is simply proxied between the browser and the node. Currently only clearnet is supported, but I believe that the WebSocket proxy code could be modified to create a socket connection to a TOR only node to make this work.
 
-## Current Limitations
+## TOR
 
-- For Commando calls, `lnmessage` will handle matching requests with responses and in most cases this just works. For RPC calls where a large response is expected (`listinvoices`, `listpays` etc) it is recommended to `await` these calls without making other calls simultaneously. A proxy server socket connection may split the response in to multiple parts. This leads to the messages possibly getting scrambled if multiple requests are made at the same time.
-- Most connections will need to be made via a WebSocket proxy server. See the WebSocket proxy section.
-- Clearnet only. I am pretty sure that this will not work out of the box with TOR connections, but I still need to try it in a TOR browser to see if it works.
+Connecting to a node over TOR requires the proxy server to support it.
 
 ## Running Locally
 

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "lnmessage",
-  "version": "0.0.19",
+  "version": "0.0.19-0.1.0",
   "description": "Talk to Lightning nodes from your browser",
   "main": "dist/index.js",
   "type": "module",
@@ -14,8 +14,9 @@
     "build": "tsc"
   },
   "devDependencies": {
-    "@types/crypto-js": "^4.1.1",
+    "@types/node": "^18.14.0",
     "@types/secp256k1": "^4.0.3",
+    "@types/ws": "^8.5.4",
     "@typescript-eslint/eslint-plugin": "^5.27.0",
     "@typescript-eslint/parser": "^5.27.0",
     "eslint": "^8.16.0",
@@ -24,9 +25,10 @@
     "typescript": "^4.8.2"
   },
   "dependencies": {
+    "@noble/hashes": "^1.2.0",
     "buffer": "^6.0.3",
-    "crypto-js": "^4.1.1",
     "rxjs": "^7.5.7",
-    "secp256k1": "^4.0.3"
+    "secp256k1": "^5.0.0",
+    "ws": "^8.12.1"
   }
 }

--- a/src/chacha/chacha20.ts
+++ b/src/chacha/chacha20.ts
@@ -37,7 +37,7 @@ class Chacha20 {
 
     this.cachePos = 64
     this.buffer = new Uint32Array(16)
-    this.output = new Buffer(64)
+    this.output = Buffer.alloc(64)
   }
 
   quarterRound(a: number, b: number, c: number, d: number) {
@@ -91,7 +91,7 @@ class Chacha20 {
 
   getBytes(len: number) {
     let dpos = 0
-    const dst = new Buffer(len)
+    const dst = Buffer.alloc(len)
     const cacheLen = 64 - this.cachePos
 
     if (cacheLen) {

--- a/src/chacha/index.ts
+++ b/src/chacha/index.ts
@@ -27,7 +27,7 @@ class Cipher {
     }
     this.alen = aad.length
     this.poly.update(aad)
-    const padding = new Buffer(padAmount(this.alen))
+    const padding = Buffer.alloc(padAmount(this.alen))
     if (padding.length) {
       padding.fill(0)
       this.poly.update(padding)
@@ -83,14 +83,14 @@ class Cipher {
       throw new Error('Unsupported state or unable to authenticate data')
     }
 
-    const padding = new Buffer(padAmount(this.clen))
+    const padding = Buffer.alloc(padAmount(this.clen))
 
     if (padding.length) {
       padding.fill(0)
       this.poly.update(padding)
     }
 
-    const lens = new Buffer(16)
+    const lens = Buffer.alloc(16)
     lens.fill(0)
     lens.writeUInt32LE(this.alen, 0)
     lens.writeUInt32LE(this.clen, 8)
@@ -110,7 +110,7 @@ class Cipher {
 
   getAuthTag() {
     if (this._decrypt || this.tag === null) {
-      return new Buffer('')
+      return Buffer.from('')
     }
     return this.tag
   }

--- a/src/chacha/poly1305.ts
+++ b/src/chacha/poly1305.ts
@@ -9,7 +9,7 @@ class Poly1305 {
   public finished: number
 
   constructor(key: Buffer) {
-    this.buffer = new Buffer(16)
+    this.buffer = Buffer.alloc(16)
     this.leftover = 0
     this.r = new Uint16Array(10)
     this.h = new Uint16Array(10)
@@ -126,7 +126,7 @@ class Poly1305 {
   }
 
   finish() {
-    let mac = new Buffer(16),
+    let mac = Buffer.alloc(16),
       g = new Uint16Array(10),
       c = 0,
       mask = 0,

--- a/src/crypto.ts
+++ b/src/crypto.ts
@@ -1,24 +1,19 @@
-import CryptoJS from 'crypto-js'
 import { Buffer } from 'buffer'
 import secp256k1 from 'secp256k1'
 import { createCipher, createDecipher } from './chacha/index.js'
+import { hmac } from '@noble/hashes/hmac'
+import { sha256 as sha256Array } from '@noble/hashes/sha256'
+import { bytesToHex, randomBytes } from '@noble/hashes/utils'
+
+export function sha256(input: Buffer): Buffer {
+  return Buffer.from(sha256Array(input))
+}
 
 export function ecdh(pubkey: Uint8Array, privkey: Uint8Array) {
   return Buffer.from(secp256k1.ecdh(pubkey, privkey))
 }
-
 export function hmacHash(key: Buffer, input: Buffer) {
-  const words = CryptoJS.HmacSHA256(
-    CryptoJS.enc.Hex.parse(input.toString('hex')),
-    CryptoJS.enc.Hex.parse(key.toString('hex'))
-  )
-
-  return Buffer.from(CryptoJS.enc.Hex.stringify(words), 'hex')
-}
-
-export async function sha256(input: Buffer): Promise<Buffer> {
-  const res = await crypto.subtle.digest('SHA-256', input)
-  return Buffer.from(res)
+  return Buffer.from(hmac(sha256Array, key, input))
 }
 
 export function hkdf(ikm: Buffer, len: number, salt = Buffer.alloc(0), info = Buffer.alloc(0)) {
@@ -99,11 +94,10 @@ export function ccpDecrypt(k: Buffer, n: Buffer, ad: Buffer, ciphertext: Buffer)
 export function createRandomPrivateKey(): string {
   let privKey
   do {
-    const bytes = Buffer.allocUnsafe(32)
-    privKey = crypto.getRandomValues(bytes)
-  } while (!validPrivateKey(privKey))
+    privKey = randomBytes(32)
+  } while (!validPrivateKey(Buffer.from(privKey)))
 
-  return privKey.toString('hex')
+  return bytesToHex(privKey)
 }
 
 export function validPublicKey(publicKey: string): boolean {
@@ -114,4 +108,8 @@ export function validPrivateKey(privateKey: string | Buffer): boolean {
   return secp256k1.privateKeyVerify(
     typeof privateKey === 'string' ? Buffer.from(privateKey, 'hex') : privateKey
   )
+}
+
+export function createRandomBytes(length: number) {
+  return randomBytes(length)
 }

--- a/src/index.ts
+++ b/src/index.ts
@@ -397,7 +397,7 @@ class LnMessage {
               this._partialCommandoMsgs[requestId],
               message.subarray(0, message.byteLength - 16)
             ])
-          : Buffer.from(decrypted.subarray(0, decrypted.length - 16))
+          : decrypted.subarray(0, decrypted.length - 16)
 
         return
       }

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,7 +1,8 @@
 import { BehaviorSubject, firstValueFrom, Observable, Subject } from 'rxjs'
 import { filter, map } from 'rxjs/operators'
 import { Buffer } from 'buffer'
-import { createRandomPrivateKey } from './crypto.js'
+import { bytesToHex } from '@noble/hashes/utils'
+import { createRandomBytes, createRandomPrivateKey } from './crypto.js'
 import { NoiseState } from './noise-state.js'
 import { validateInit } from './validation.js'
 import { deserialize } from './messages/MessageFactory.js'
@@ -10,6 +11,7 @@ import { BufferReader, BufferWriter } from './messages/buf.js'
 import { CommandoMessage } from './messages/CommandoMessage.js'
 import { PongMessage } from './messages/PongMessage.js'
 import { PingMessage } from './messages/PingMessage.js'
+import type { WebSocket as NodeWebSocket } from 'ws'
 
 import {
   LnWebSocketOptions,
@@ -45,7 +47,7 @@ class LnMessage {
    */
   public wsUrl: string
   /**The WebSocket instance*/
-  public socket: WebSocket | null
+  public socket: WebSocket | NodeWebSocket | null
   /**
    * @deprecated Use connectionStatus$ instead
    */
@@ -68,7 +70,7 @@ class LnMessage {
     (JsonRpcSuccessResponse | JsonRpcErrorResponse) & { reqId: string }
   >
   /**Node JS Buffer instance, useful if handling decrypted messages manually*/
-  public Buffer: BufferConstructor
+  public Buffer: Buffer['constructor']
 
   private _ls: Buffer
   private _es: Buffer
@@ -156,14 +158,17 @@ class LnMessage {
     this.connecting = true
     this.connectionStatus$.next('connecting')
     this._attemptReconnect = attemptReconnect
-    this.socket = new WebSocket(this.wsUrl)
+
+    this.socket = new (
+      typeof window === 'undefined' ? (await import('ws')).default : window.WebSocket
+    )(this.wsUrl)
     this.socket.binaryType = 'arraybuffer'
 
     this.socket.onopen = async () => {
       this._log('info', 'WebSocket is connected')
       this._log('info', 'Creating Act1 message')
 
-      const msg = await this.noise.initiatorAct1(Buffer.from(this.remoteNodePublicKey, 'hex'))
+      const msg = this.noise.initiatorAct1(Buffer.from(this.remoteNodePublicKey, 'hex'))
 
       if (this.socket) {
         this._log('info', 'Sending Act1 message')
@@ -192,7 +197,7 @@ class LnMessage {
       }
     }
 
-    this.socket.onerror = (err) => {
+    this.socket.onerror = (err: { message: string }) => {
       this._log('error', `WebSocket error: ${JSON.stringify(err)}`)
     }
 
@@ -250,16 +255,16 @@ class LnMessage {
 
             // Initiator Act2
             case HANDSHAKE_STATE.AWAITING_RESPONDER_REPLY:
-              readMore = await this._processResponderReply()
+              readMore = this._processResponderReply()
               break
           }
         } else {
           switch (this._readState) {
             case READ_STATE.READY_FOR_LEN:
-              readMore = await this._processPacketLength()
+              readMore = this._processPacketLength()
               break
             case READ_STATE.READY_FOR_BODY:
-              readMore = await this._processPacketBody()
+              readMore = this._processPacketBody()
               break
             case READ_STATE.BLOCKED:
               readMore = false
@@ -279,7 +284,7 @@ class LnMessage {
     this._processingBuffer = false
   }
 
-  private async _processResponderReply() {
+  private _processResponderReply() {
     // read 50 bytes
     const data = this._messageBuffer.readBytes(50)
 
@@ -289,11 +294,11 @@ class LnMessage {
 
     // process reply
     this._log('info', 'Validating message as part of Act2')
-    await this.noise.initiatorAct2(data)
+    this.noise.initiatorAct2(data)
 
     // create final act of the handshake
     this._log('info', 'Creating reply for Act3')
-    const reply = await this.noise.initiatorAct3()
+    const reply = this.noise.initiatorAct3()
 
     if (this.socket) {
       this._log('info', 'Sending reply for act3')
@@ -308,7 +313,7 @@ class LnMessage {
     return true
   }
 
-  private async _processPacketLength() {
+  private _processPacketLength() {
     const LEN_CIPHER_BYTES = 2
     const LEN_MAC_BYTES = 16
 
@@ -320,7 +325,7 @@ class LnMessage {
       if (!lc) return false
 
       // Decrypt the length including the MAC
-      const l = await this.noise.decryptLength(lc)
+      const l = this.noise.decryptLength(lc)
 
       // We need to store the value in a local variable in case
       // we are unable to read the message body in its entirety.
@@ -338,7 +343,7 @@ class LnMessage {
     }
   }
 
-  private async _processPacketBody() {
+  private _processPacketBody() {
     const MESSAGE_MAC_BYTES = 16
 
     if (!this._l) return false
@@ -353,7 +358,7 @@ class LnMessage {
       if (!c) return false
 
       // Decrypt the full message cipher + MAC
-      const m = await this.noise.decryptMessage(c)
+      const m = this.noise.decryptMessage(c)
 
       // Now that we've read the message, we can remove the
       // cached length before we transition states
@@ -392,7 +397,7 @@ class LnMessage {
               this._partialCommandoMsgs[requestId],
               message.subarray(0, message.byteLength - 16)
             ])
-          : decrypted.subarray(0, decrypted.length - 16)
+          : Buffer.from(decrypted.subarray(0, decrypted.length - 16))
 
         return
       }
@@ -415,7 +420,7 @@ class LnMessage {
       switch (payload.type) {
         case MessageType.Init: {
           this._log('info', 'Constructing Init message reply')
-          const reply = await this.noise.encryptMessage((payload as IWireMessage).serialize())
+          const reply = this.noise.encryptMessage((payload as IWireMessage).serialize())
 
           if (this.socket) {
             this._log('info', 'Sending Init message reply')
@@ -437,7 +442,7 @@ class LnMessage {
           this._log('info', 'Creating a Pong message')
 
           const pongMessage = new PongMessage((payload as PingMessage).numPongBytes).serialize()
-          const pong = await this.noise.encryptMessage(pongMessage)
+          const pong = this.noise.encryptMessage(pongMessage)
 
           if (this.socket) {
             this._log('info', 'Sending a Pong message')
@@ -489,9 +494,8 @@ class LnMessage {
 
     if (!reqId) {
       // create random id to match request with response
-      const idBytes = Buffer.allocUnsafe(8)
-      const id = crypto.getRandomValues(idBytes)
-      reqId = id.toString('hex')
+      const id = createRandomBytes(8)
+      reqId = bytesToHex(id)
     }
 
     // write the type
@@ -512,7 +516,7 @@ class LnMessage {
     )
 
     this._log('info', 'Creating message to send')
-    const message = await this.noise.encryptMessage(writer.toBuffer())
+    const message = this.noise.encryptMessage(writer.toBuffer())
 
     if (this.socket) {
       this._log('info', 'Sending commando message')

--- a/src/noise-state.ts
+++ b/src/noise-state.ts
@@ -1,6 +1,6 @@
+import { Buffer } from 'buffer'
 import { ccpDecrypt, ccpEncrypt, ecdh, getPublicKey, hkdf, sha256 } from './crypto.js'
 import type { NoiseStateOptions } from './types.js'
-import { Buffer } from 'buffer'
 
 export class NoiseState {
   /**
@@ -107,12 +107,12 @@ export class NoiseState {
    * @param rpk remote public key
    * @return Buffer that is 50 bytes
    */
-  public async initiatorAct1(rpk: Buffer): Promise<Buffer> {
+  public initiatorAct1(rpk: Buffer): Buffer {
     this.rpk = rpk
-    await this._initialize(this.rpk)
+    this._initialize(this.rpk)
 
     // 2. h = SHA-256(h || epk)
-    this.h = await sha256(Buffer.concat([this.h, this.epk]))
+    this.h = sha256(Buffer.concat([this.h, this.epk]))
 
     // 3. es = ECDH(e.priv, rs)
     const ss = ecdh(this.rpk, this.es)
@@ -126,7 +126,7 @@ export class NoiseState {
     const c = ccpEncrypt(this.tempK1, Buffer.alloc(12), this.h, Buffer.alloc(0))
 
     // 6. h = SHA-256(h || c)
-    this.h = await sha256(Buffer.concat([this.h, c]))
+    this.h = sha256(Buffer.concat([this.h, c]))
     // 7. m = 0 || epk || c
     const m = Buffer.concat([Buffer.alloc(1), this.epk, c])
     return m
@@ -138,7 +138,7 @@ export class NoiseState {
    *
    * @param m 50-byte message from responder's act1
    */
-  public async initiatorAct2(m: Buffer) {
+  public initiatorAct2(m: Buffer) {
     // 1. read exactly 50 bytes off the stream
     if (m.length !== 50) throw new Error('ACT2_READ_FAILED')
     // 2. parse th read message m into v, re, and c
@@ -150,7 +150,7 @@ export class NoiseState {
     // 3. assert version is known version
     if (v !== 0) throw new Error('ACT2_BAD_VERSION')
     // 4. sha256(h || re.serializedCompressed');
-    this.h = await sha256(Buffer.concat([this.h, this.repk]))
+    this.h = sha256(Buffer.concat([this.h, this.repk]))
     // 5. ss = ECDH(re, e.priv);
     const ss = ecdh(this.repk, this.es)
     // 6. ck, temp_k2 = HKDF(cd, ss)
@@ -160,7 +160,7 @@ export class NoiseState {
     // 7. p = decryptWithAD()
     ccpDecrypt(this.tempK2, Buffer.alloc(12), this.h, c)
     // 8. h = sha256(h || c)
-    this.h = await sha256(Buffer.concat([this.h, c]))
+    this.h = sha256(Buffer.concat([this.h, c]))
   }
   /**
    * Initiator Act3 is the final phase in the authenticated
@@ -168,7 +168,7 @@ export class NoiseState {
    * was successful. The initiator transports its static public key
    * to the responder.
    */
-  public async initiatorAct3() {
+  public initiatorAct3() {
     // 1. c = encryptWithAD(temp_k2, 1, h, lpk)
     const c = ccpEncrypt(
       this.tempK2,
@@ -177,7 +177,7 @@ export class NoiseState {
       this.lpk
     )
     // 2. h = sha256(h || c)
-    this.h = await sha256(Buffer.concat([this.h, c]))
+    this.h = sha256(Buffer.concat([this.h, c]))
     // 3. ss = ECDH(re, s.priv)
     const ss = ecdh(this.repk, this.ls)
     // 4. ck, temp_k3 = HKDF(ck, ss)
@@ -202,8 +202,8 @@ export class NoiseState {
    * validates that the initiator knows the receivers public key.
    * @param m 50-byte message sent by the initiator
    */
-  public async receiveAct1(m: Buffer) {
-    await this._initialize(this.lpk)
+  public receiveAct1(m: Buffer) {
+    this._initialize(this.lpk)
     // 1. read exactly 50 bytes off the stream
     if (m.length !== 50) throw new Error('ACT1_READ_FAILED')
     // 2. parse th read message m into v,re, and c
@@ -214,7 +214,7 @@ export class NoiseState {
     // 3. assert version is known version
     if (v !== 0) throw new Error('ACT1_BAD_VERSION')
     // 4. sha256(h || re.serializedCompressed');
-    this.h = await sha256(Buffer.concat([this.h, re]))
+    this.h = sha256(Buffer.concat([this.h, re]))
     // 5. ss = ECDH(re, ls.priv);
     const ss = ecdh(re, this.ls)
     // 6. ck, temp_k1 = HKDF(cd, ss)
@@ -224,16 +224,16 @@ export class NoiseState {
     // 7. p = decryptWithAD(temp_k1, 0, h, c)
     ccpDecrypt(this.tempK1, Buffer.alloc(12), this.h, c)
     // 8. h = sha256(h || c)
-    this.h = await sha256(Buffer.concat([this.h, c]))
+    this.h = sha256(Buffer.concat([this.h, c]))
   }
   /**
    * Receiver Act2 takes place only if Act1 was successful.
    * This act sends responder's ephermeral key to the initiator.
    */
-  public async recieveAct2(): Promise<Buffer> {
+  public recieveAct2(): Buffer {
     // 1. e = generateKey() => done in initialization
     // 2. h = sha256(h || e.pub.compressed())
-    this.h = await sha256(Buffer.concat([this.h, this.epk]))
+    this.h = sha256(Buffer.concat([this.h, this.epk]))
     // 3. ss = ecdh(re, e.priv)
     const ss = ecdh(this.repk, this.es)
     // 4. ck, temp_k2 = hkdf(ck, ss)
@@ -243,7 +243,7 @@ export class NoiseState {
     // 5. c = encryptWithAd(temp_k2, 0, h, zero)
     const c = ccpEncrypt(this.tempK2, Buffer.alloc(12), this.h, Buffer.alloc(0))
     // 6. h = sha256(h || c)
-    this.h = await sha256(Buffer.concat([this.h, c]))
+    this.h = sha256(Buffer.concat([this.h, c]))
     // 7. m = 0 || e.pub.compressed() Z|| c
     const m = Buffer.concat([Buffer.alloc(1), this.epk, c])
     return m
@@ -254,7 +254,7 @@ export class NoiseState {
    * The receiver extracts the public key of the initiator.
    * @param m 66-byte message
    */
-  public async receiveAct3(m: Buffer) {
+  public receiveAct3(m: Buffer) {
     // 1. read exactly 66 bytes from the network buffer
     if (m.length !== 66) throw new Error('ACT3_READ_FAILED')
     // 2. parse m into v, c, t
@@ -267,7 +267,7 @@ export class NoiseState {
     const rs = ccpDecrypt(this.tempK2, Buffer.from('000000000100000000000000', 'hex'), this.h, c)
     this.rpk = rs as Buffer
     // 5. h = sha256(h || c)
-    this.h = await sha256(Buffer.concat([this.h, c]))
+    this.h = sha256(Buffer.concat([this.h, c]))
     // 6. ss = ECDH(rs, e.priv)
     const ss = ecdh(this.rpk, this.es)
     // 7. ck, temp_k3 = hkdf(cs, ss)
@@ -290,7 +290,7 @@ export class NoiseState {
    * rotated every 1000 messages.
    * @param m
    */
-  public async encryptMessage(m: Buffer): Promise<Buffer> {
+  public encryptMessage(m: Buffer): Buffer {
     // step 1/2. serialize m length into int16
     const l = Buffer.alloc(2)
     l.writeUInt16BE(m.length, 0)
@@ -309,7 +309,7 @@ export class NoiseState {
    * Decrypts the length of the message using the receiving key and nonce.
    * The receiving key is rotated every 1000 messages.
    */
-  public async decryptLength(lc: Buffer): Promise<number> {
+  public decryptLength(lc: Buffer): number {
     const l = ccpDecrypt(this.rk, this.rn, Buffer.alloc(0), lc) as Buffer
     if (this._incrementRecievingNonce() >= 1000) this._rotateRecievingKeys()
     return l.readUInt16BE(0)
@@ -318,7 +318,7 @@ export class NoiseState {
    * Decrypts the message using the receiving key and nonce. The receiving
    * key is rotated every 1000 messages.
    */
-  public async decryptMessage(c: Buffer) {
+  public decryptMessage(c: Buffer) {
     const m = ccpDecrypt(this.rk, this.rn, Buffer.alloc(0), c)
     if (this._incrementRecievingNonce() >= 1000) this._rotateRecievingKeys()
     return m as Buffer
@@ -327,18 +327,18 @@ export class NoiseState {
   /**
    * Initializes the noise state prior to Act1.
    */
-  private async _initialize(pubkey: Buffer) {
+  private _initialize(pubkey: Buffer) {
     // 1. h = SHA-256(protocolName)
-    this.h = await sha256(Buffer.from(this.protocolName))
+    this.h = sha256(Buffer.from(this.protocolName))
 
     // 2. ck = h
     this.ck = this.h
 
     // 3. h = SHA-256(h || prologue)
-    this.h = await sha256(Buffer.concat([this.h, this.prologue]))
+    this.h = sha256(Buffer.concat([this.h, this.prologue]))
 
     // 4. h = SHA-256(h || pubkey)
-    this.h = await sha256(Buffer.concat([this.h, pubkey]))
+    this.h = sha256(Buffer.concat([this.h, pubkey]))
   }
 
   private _incrementSendingNonce() {

--- a/src/types.ts
+++ b/src/types.ts
@@ -1,3 +1,5 @@
+import type { Buffer } from 'buffer'
+
 export type LnWebSocketOptions = {
   /**
    * 33-byte hex remote compressed public key.

